graph_coloring.py

```diff
@@ -12,78 +12,107 @@
-sampleset = sampler.sample_dqm(dqm, label='Example - Graph Coloring')
-
-# get the first solution, and print it
-sample = sampleset.first.sample
-node_colors = [sample[i] for i in G.nodes()]
-nx.draw(G, pos=pos, node_color=node_colors, node_size=50, edgecolors='k', cmap='hsv')
-plt.savefig('graph_result.png')
-
-# check that colors are different
-valid = True
-for edge in G.edges:
-    i, j = edge
-    if sample[i] == sample[j]:
-        valid = False
-        break
-print("\nSolution validity: ", valid)
-
-colors_used = max(sample.values())+1
-print("\nColors required:", colors_used, "\n")
+def build_graph(num_nodes):
+    """Build graph."""
+
+    print("\nBuilding graph...")
+
+    G = nx.powerlaw_cluster_graph(num_nodes, 3, 0.4)
+    pos = nx.spring_layout(G)
+    nx.draw(G, pos=pos, node_size=50, edgecolors='k', cmap='hsv')
+    plt.savefig("original_graph.png")
+
+    return G, pos
+
+def build_cqm(G, num_colors):
+    """Build CQM model."""
+
+    print("\nBuilding constrained quadratic model...")
+
+    # Initialize the CQM object
+    cqm = ConstrainedQuadraticModel()
+
+    # Build CQM variables
+    colors = {n: {c: Binary((n, c)) for c in range(num_colors)} for n in G.nodes}
+
+    # Add constraint to make variables discrete
+    for n in G.nodes():
+        cqm.add_discrete([(n, c) for c in range(num_colors)])
+  
+    # Build the constraints: edges have different color end points
+    for u, v in G.edges:
+        for c in range(num_colors):
+            cqm.add_constraint(colors[u][c]*colors[v][c] == 0)
+
+    return cqm
+
+def run_hybrid_solver(cqm):
+    """Solve CQM using hybrid solver."""
+
+    print("\nRunning hybrid sampler...")
+
+    # Initialize the CQM solver
+    sampler = LeapHybridCQMSampler()
+
+    # Solve the problem using the CQM solver
+    sampleset = sampler.sample_cqm(cqm, label='Example - Graph Coloring')
+    feasible_sampleset = sampleset.filter(lambda row: row.is_feasible)
+
+    try:
+        sample = feasible_sampleset.first.sample
+    except:
+        print("\nNo feasible solutions found.")
+        exit()
+
+    soln = {key[0]: key[1] for key, val in sample.items() if val == 1.0}
+
+    return soln
+
+def plot_soln(sample, pos):
+    """Plot results and save file.
+    
+    Args:
+        sample (dict):
+            Sample containing a solution. Each key is a node and each value 
+            is an int representing the node's color.
+
+        pos (dict):
+            Plotting information for graph so that same graph shape is used.
+    """
+
+    print("\nProcessing sample...")
+
+    node_colors = [sample[i] for i in G.nodes()]
+    nx.draw(G, pos=pos, node_color=node_colors, node_size=50, edgecolors='k', cmap='hsv')
+    fname = 'graph_result.png'
+    plt.savefig(fname)
+
+    print("\nSaving results in {}...".format(fname))
+
+# ------- Main program -------
+if __name__ == "__main__":
+
+    num_nodes = 50
+
+    G, pos = build_graph(num_nodes)
+    num_colors = max(d for _, d in G.degree())-1
```

Comment:
Should this be degree + 1?
```suggestion
    num_colors = max(d for _, d in G.degree()) + 1
```